javascript/computer_science/project_hash_map.md

```diff
@@ -0,0 +1,223 @@
+### Introduction
+
+One of the most used data structure across all languages is a Hash Table, aka Hash Map. If you worked before with JavaScript Object Literal `{}`, `Set`, and `Map` then you are already familiar with a hash table. But how do they work exactly, how can we save values using strings and symbols as keys and then retrieving those values using the same keys.
+
+In this project, you will learn how it all works, and you will implement your own Hash Map! To start we need to learn a few concepts, one of which is a hash code, what does it mean, and how to use it. Let us get started and hope you are exited to finally know what is happening with an object when you put keys and values in them.
+
+### What is a hash code
+
+Let us start by learning what does it mean to hash a value. Hashing means taking an input in and generating an output that represents this input. A Hash function should be a pure function, hashing the same input should always return the same hash code, there should be no random generation that is happening. For example, let's work a hashing function that takes a name and gives us the first letter of that name:
+
+```javascript
+function hash(name) {
+  return name.charAt(0);
+}
+```
+
+We created our first basic hashing function, it takes a name and returns the first letter of that name.
+You might be familiar with this concept from your cipher exercise earlier in foundation, but there is a key difference between hashing and ciphering (encryption), which is reversibility.
+
+Hashing is a one-way process. Here is an example given a name, you can get a hash out of it, but given a hash, you cannot revert it back to a name. If you have a name `"Carlos"` we can hash it to `"C"`. But it will not be possible to reverse it from `"C"` back to its original form. You cannot know if it is `"Carlos"`, maybe it's `"Carla"` or `"Carrot"` we don't know. This is very good for security also, given a password you can save the hash of that password, but if someone steals your hashes they cannot know the original passwords since they are unable to reverse the hash back to the password. We are not going to dive deeper into hashing passwords, this is a subject for another lesson.
```

Comment:
```suggestion
Hashing is a one-way process. In the above example, you can get a hash code from a name, but when given a hash, you cannot revert it back to a name. If you have a name `"Carlos"` we can hash it to `"C"`. But it will not be possible to reverse it from `"C"` back to its original form. You cannot know if the original name was `"Carlos"`. Maybe it's `"Carla"` or `"Carrot"`. We can't know. This is very good for security. Given a password you can save the hash of that password, but if someone steals your hashes they cannot know the original passwords because they are unable to reverse the hash back to the password. We are not going to dive deeper into hashing passwords, and this is a subject for another lesson.
```
I also think the information about passwords might fit better as a note.